src/sage/graphs/generators/random.py

```diff
@@ -1459,6 +1459,122 @@ def RandomTreePowerlaw(n, gamma=3, tries=1000, seed=None):
+    r"""
+    Return a random `k`-tree on `n` nodes numbered `0` through `n-1`.
+
+    ALGORITHM:
+
+    The algorithm first generates a complete graph on `k + 1` vertices.
+    Vertices are subsequently generated by randomly choosing one of the
+    existing cliques in the graph, and creating a new clique by replacing
+    one of the vertices in the selected clique with a newly created one.
+
+    INPUT:
+
+    - ``n`` -- number of vertices in the `k`-tree
+
+    - ``k`` -- within a clique each vertex is connected to `k` vertices. `k`
+      also corresponds to the treewidth of the `k`-tree
+
+    - ``seed`` -- a ``random.Random`` seed or a Python ``int`` for the random
+      number generator (default: ``None``)
+
+    TESTS::
+
+        sage: g=graphs.RandomKTree(50,5)
+        sage: g.size()
+        235
+        sage: g.order()
+        50
+        sage: g.treewidth()
+        5
+
+    EXAMPLES::
+
+        sage: G = graphs.RandomKTree(50, 5)
+        sage: G.treewidth()
+        5
+        sage: G.show()  # not tested
+    """
+    if n < k + 1:
+        raise ValueError("n must be greater than k + 1")
+
+    if seed is not None:
+        set_random_seed(seed)
+
+    g = Graph(name=f"Random {k}-tree of order {n}")
+    g.add_clique(list(range(k + 1)))
+
+    cliques = [list(range(k+1))]
+
+    # Randomly choose a row, and copy 1 of the cliques
+    # One of those vertices is then replaced with a new vertex
+    for newVertex in range(k + 1, n):
+        copiedClique = cliques[randint(0, len(cliques)-1)].copy()
+        copiedClique[randint(0, k)] = newVertex
+        cliques.append(copiedClique)
+        for u in copiedClique:
+            if u != newVertex:
+                g.add_edge(u, newVertex)
+    return g
+
+
+def RandomPartialKTree(n, k, x, seed=None):
+    r"""
+    Return a random partial `k`-tree on `n` nodes.
+
+    A partial `k`-tree is defined as a subgraph of a `k`-tree. This can also be
+    described as a graph with treewidth at most `k`.
+
+    INPUT:
+
+    - ``n`` -- number of vertices in the `k`-tree
+
+    - ``k`` -- within a clique each vertex is connected to `k` vertices. `k`
+      also corresponds to the treewidth of the `k`-tree
+
+    - ``x`` -- how many edges are deleted from the `k`-tree
+
+    - ``seed`` -- a ``random.Random`` seed or a Python ``int`` for the random
+      number generator (default: ``None``)
+
+    TESTS::
+
+        sage: g=graphs.RandomPartialKTree(50,5,2)
+        sage: g.order()
+        50
+        sage: g.size()
+        233
+        sage: g.treewidth()
+        5
+
+    EXAMPLES::
+
+        sage: G = graphs.RandomPartialKTree(50,5,2)
+        sage: G.treewidth()
+        5
+        sage: G.show()  # not tested
+    """
+    if seed is not None:
+        set_random_seed(seed)
+
+    g = RandomKTree(n, k, seed)
+
+    # Check that x doesn't delete too many edges
+    # This formula calculates how many edges are in `k`-tree with `n` nodes
+    if x > (k ^ 2 + k) / 2 + (n - k - 1) * k:
```

Comment:
In case of equality, the graph has no edges. May be we can avoid the call to `RandomKTree` in this case ?